{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d75f6dfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: sagemaker>=2.99.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (2.103.0)\n",
      "Requirement already satisfied: protobuf<4.0,>=3.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (3.15.2)\n",
      "Requirement already satisfied: boto3<2.0,>=1.20.21 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (1.21.42)\n",
      "Requirement already satisfied: protobuf3-to-dict<1.0,>=0.1.5 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (0.1.5)\n",
      "Requirement already satisfied: numpy<2.0,>=1.9.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (1.19.5)\n",
      "Requirement already satisfied: attrs<22,>=20.3.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (20.3.0)\n",
      "Requirement already satisfied: pandas in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (1.1.5)\n",
      "Requirement already satisfied: pathos in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (0.2.8)\n",
      "Requirement already satisfied: smdebug-rulesconfig==1.0.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (1.0.1)\n",
      "Requirement already satisfied: importlib-metadata<5.0,>=1.4.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (3.7.0)\n",
      "Requirement already satisfied: google-pasta in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (0.2.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from sagemaker>=2.99.0) (21.3)\n",
      "Requirement already satisfied: s3transfer<0.6.0,>=0.5.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from boto3<2.0,>=1.20.21->sagemaker>=2.99.0) (0.5.0)\n",
      "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from boto3<2.0,>=1.20.21->sagemaker>=2.99.0) (0.10.0)\n",
      "Requirement already satisfied: botocore<1.25.0,>=1.24.42 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from boto3<2.0,>=1.20.21->sagemaker>=2.99.0) (1.24.42)\n",
      "Requirement already satisfied: typing-extensions>=3.6.4 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from importlib-metadata<5.0,>=1.4.0->sagemaker>=2.99.0) (4.0.1)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from importlib-metadata<5.0,>=1.4.0->sagemaker>=2.99.0) (3.4.0)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from packaging>=20.0->sagemaker>=2.99.0) (2.4.7)\n",
      "Requirement already satisfied: six>=1.9 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from protobuf<4.0,>=3.1->sagemaker>=2.99.0) (1.15.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from pandas->sagemaker>=2.99.0) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from pandas->sagemaker>=2.99.0) (2021.1)\n",
      "Requirement already satisfied: dill>=0.3.4 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from pathos->sagemaker>=2.99.0) (0.3.4)\n",
      "Requirement already satisfied: ppft>=1.6.6.4 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from pathos->sagemaker>=2.99.0) (1.6.6.4)\n",
      "Requirement already satisfied: pox>=0.3.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from pathos->sagemaker>=2.99.0) (0.3.0)\n",
      "Requirement already satisfied: multiprocess>=0.70.12 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from pathos->sagemaker>=2.99.0) (0.70.12.2)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.25.4 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from botocore<1.25.0,>=1.24.42->boto3<2.0,>=1.20.21->sagemaker>=2.99.0) (1.26.8)\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "!{sys.executable} -m pip install \"sagemaker>=2.99.0\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b054a6ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sagemaker\n",
    "from sagemaker.local import LocalSession\n",
    "from sagemaker import get_execution_role\n",
    "from sagemaker.session import Session\n",
    "import boto3\n",
    "from sagemaker.sklearn.estimator import SKLearn\n",
    "from sagemaker.workflow.pipeline_context import PipelineSession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ef2d2497",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training input/output will be stored in sagemaker-us-east-1-485822383573/chapter9/data\n",
      "\n",
      "IAM Role: arn:aws:iam::485822383573:role/service-role/AmazonSageMaker-ExecutionRole-20220426T122295\n"
     ]
    }
   ],
   "source": [
    "sagemaker_session=sagemaker.Session()\n",
    "local_session=LocalSession()\n",
    "pipeline_session = PipelineSession()\n",
    "bucket = sagemaker_session.default_bucket()\n",
    "\n",
    "role = get_execution_role()\n",
    "prefix = 'chapter9/data'\n",
    "\n",
    "print('Training input/output will be stored in {}/{}'.format(bucket, prefix))\n",
    "print('\\nIAM Role: {}'.format(role))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6b855a8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting scripts/preprocessing.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile scripts/preprocessing.py\n",
    "import csv\n",
    "import wget\n",
    "import zipfile\n",
    "import os\n",
    "import pandas as pd\n",
    "import boto3\n",
    "import time\n",
    "import json\n",
    "import argparse\n",
    "from sklearn.preprocessing import Normalizer\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "import numpy as np\n",
    "\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument('--bucket', type=str)\n",
    "parser.add_argument('--region',type=str)\n",
    "args = parser.parse_args()\n",
    "\n",
    "bucket=args.bucket\n",
    "cm = boto3.client('comprehendmedical',region_name=args.region)\n",
    "s3_client = boto3.client('s3',region_name=args.region)\n",
    "\n",
    "if os.path.exists('data')==False:\n",
    "    os.mkdir('data')\n",
    "\n",
    "file_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/00462/drugsCom_raw.zip'\n",
    "dest_file = 'data/drugsCom_raw.zip'\n",
    "\n",
    "print(\"Downloading source files...\")\n",
    "\n",
    "wget.download(file_url, dest_file)\n",
    "\n",
    "with zipfile.ZipFile('data/drugsCom_raw.zip', 'r') as zip_ref:\n",
    "    zip_ref.extractall('data')\n",
    "\n",
    "os.remove('data/drugsCom_raw.zip')\n",
    "\n",
    "orig_list = list()\n",
    "for filename in os.listdir('data'):\n",
    "    with open('data/'+filename) as csvfile:\n",
    "        myreader = csv.reader(csvfile, delimiter='\\t')\n",
    "        for row in myreader:\n",
    "            if row[0] == '':\n",
    "                continue\n",
    "            else:\n",
    "                orig_list.append({\n",
    "                    'id': row[0],\n",
    "                    'drugName': row[1],\n",
    "                    'condition': row[2],\n",
    "                    'review': row[3]\n",
    "                })\n",
    "\n",
    "    \n",
    "if os.path.exists('processed_data')==False:\n",
    "    os.mkdir('processed_data')\n",
    "    \n",
    "raw_df=pd.DataFrame.from_records(orig_list)\n",
    "raw_df.to_csv('processed_data/raw_df.csv', index=False)\n",
    "\n",
    "print(\"\\nRaw data processed from input files\")\n",
    "print(\"\\nRamdomly sampling 100 rows for topic extraction\")\n",
    "\n",
    "df_sample=raw_df.sample(n=100)\n",
    "sample_list = list()\n",
    "\n",
    "\n",
    "for index,row in df_sample.iterrows():\n",
    "    entities = cm.detect_entities(Text=row['review'])\n",
    "    topic_list = []\n",
    "    for entity in entities['Entities']:\n",
    "        if entity['Category'] == 'MEDICAL_CONDITION':\n",
    "            topic_list.append(entity['Text'])\n",
    "\n",
    "    sample_list.append({\n",
    "            'id': row['id'],\n",
    "            'drugName': row['drugName'],\n",
    "            'condition': row['condition'],\n",
    "            'review': row['review'],\n",
    "            'topics': topic_list[:5]\n",
    "        })\n",
    "        \n",
    "sample_df=pd.DataFrame.from_records(sample_list)\n",
    "\n",
    "sample_df.to_csv('processed_data/sample_df.csv', index=False) \n",
    "\n",
    "\n",
    "sampled_topics=pd.read_csv('processed_data/sample_df.csv')['topics'].tolist()\n",
    "print(sampled_topics)\n",
    "vectorizer = TfidfVectorizer()\n",
    "vecs = vectorizer.fit_transform(sampled_topics)\n",
    "normalizer = Normalizer(copy=False)\n",
    "normalized_data = normalizer.fit_transform(vecs).toarray()\n",
    "normalized_data.shape\n",
    "np.savetxt(\"processed_data/prediction_data.csv\", normalized_data, delimiter=\",\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "s3_client.upload_file('processed_data/sample_df.csv', bucket, 'chapter9/data/sample_df.csv')\n",
    "s3_client.upload_file('processed_data/raw_df.csv', bucket, 'chapter9/data/raw_df.csv')\n",
    "s3_client.upload_file('processed_data/prediction_data.csv', bucket, 'chapter9/data/prediction_data.csv')\n",
    "\n",
    "\n",
    "print(\"\\nprocessed files uploaded to s3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "13a6cb29",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!python scripts/preprocessing.py --bucket $bucket --region $sagemaker_session.boto_session.region_name "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4cb03817",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "485822383573\n",
      "Login Succeeded\n",
      "\r\n",
      "Step 1/8 : FROM python:3.7-slim-buster\n",
      " ---> 8fe6e55c0412\n",
      "Step 2/8 : RUN pip install pandas\n",
      " ---> Using cache\n",
      " ---> ed3c2aadaa6e\n",
      "Step 3/8 : RUN pip install wget\n",
      " ---> Using cache\n",
      " ---> 93dc76d1c100\n",
      "Step 4/8 : RUN pip install boto3\n",
      " ---> Using cache\n",
      " ---> 43acfff1ec93\n",
      "Step 5/8 : RUN pip install sagemaker\n",
      " ---> Using cache\n",
      " ---> 0a0768240618\n",
      "Step 6/8 : RUN pip install scikit-learn\n",
      " ---> Using cache\n",
      " ---> 2ce6c8fc1e49\n",
      "Step 7/8 : ENV PYTHONUNBUFFERED=TRUE\n",
      " ---> Using cache\n",
      " ---> f68470613295\n",
      "Step 8/8 : ENTRYPOINT [\"python3\"]\n",
      " ---> Using cache\n",
      " ---> 41fbb5b0e27c\n",
      "Successfully built 41fbb5b0e27c\n",
      "Successfully tagged sagemaker-preprocessing:latest\n",
      "The push refers to repository [485822383573.dkr.ecr.us-east-1.amazonaws.com/sagemaker-preprocessing]\n",
      "20e4ef78b000: Preparing\n",
      "1cd08d11abf2: Preparing\n",
      "073fe9ab5fca: Preparing\n",
      "873a3963f49c: Preparing\n",
      "32682a294d34: Preparing\n",
      "cd77cebc5d3e: Preparing\n",
      "c899963fae46: Preparing\n",
      "353cc9dc1c96: Preparing\n",
      "c89d0deb3e29: Preparing\n",
      "735956b91a18: Preparing\n",
      "cd77cebc5d3e: Waiting\n",
      "c899963fae46: Waiting\n",
      "353cc9dc1c96: Waiting\n",
      "c89d0deb3e29: Waiting\n",
      "735956b91a18: Waiting\n",
      "873a3963f49c: Layer already exists\n",
      "20e4ef78b000: Layer already exists\n",
      "073fe9ab5fca: Layer already exists\n",
      "32682a294d34: Layer already exists\n",
      "1cd08d11abf2: Layer already exists\n",
      "cd77cebc5d3e: Layer already exists\n",
      "c899963fae46: Layer already exists\n",
      "353cc9dc1c96: Layer already exists\n",
      "735956b91a18: Layer already exists\n",
      "c89d0deb3e29: Layer already exists\n",
      "latest: digest: sha256:286e052a69cc9631505acf4a0051eb03e2229f2a616a514ea994839acda822fe size: 2428\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING! Using --password via the CLI is insecure. Use --password-stdin.\n",
      "WARNING! Your password will be stored unencrypted in /home/ec2-user/.docker/config.json.\n",
      "Configure a credential helper to remove this warning. See\n",
      "https://docs.docker.com/engine/reference/commandline/login/#credentials-store\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%sh\n",
    "\n",
    "docker_name=sagemaker-preprocessing\n",
    "account=$(aws sts get-caller-identity --query Account --output text)\n",
    "echo $account\n",
    "region=$(aws configure get region)\n",
    "\n",
    "fullname=\"${account}.dkr.ecr.${region}.amazonaws.com/${docker_name}:latest\"\n",
    "# If the repository doesn't exist in ECR, create it.\n",
    "aws ecr describe-repositories --repository-names \"${docker_name}\" > /dev/null 2>&1\n",
    "if [ $? -ne 0 ]\n",
    "then\n",
    "    aws ecr create-repository --repository-name \"${docker_name}\" > /dev/null\n",
    "fi\n",
    "\n",
    "# Get the login command from ECR and execute it directly\n",
    "$(aws ecr get-login --region ${region} --no-include-email)\n",
    "docker build -t $docker_name -f scripts/Dockerfile .\n",
    "docker tag ${docker_name} ${fullname}\n",
    "docker push ${fullname}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ceb580ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "485822383573.dkr.ecr.us-east-1.amazonaws.com/sagemaker-preprocessing:latest\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/sagemaker/workflow/pipeline_context.py:198: UserWarning: Running within a PipelineSession, there will be No Wait, No Logs, and No Job being started.\n",
      "  UserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Job Name:  sagemaker-preprocessing-2022-08-13-05-03-07-681\n",
      "Inputs:  [{'InputName': 'code', 'AppManaged': False, 'S3Input': {'S3Uri': 's3://sagemaker-us-east-1-485822383573/sagemaker-preprocessing-2022-08-13-05-03-07-681/input/code/preprocessing.py', 'LocalPath': '/opt/ml/processing/input/code', 'S3DataType': 'S3Prefix', 'S3InputMode': 'File', 'S3DataDistributionType': 'FullyReplicated', 'S3CompressionType': 'None'}}]\n",
      "Outputs:  []\n"
     ]
    }
   ],
   "source": [
    "from sagemaker.processing import ScriptProcessor\n",
    "from sagemaker.workflow.steps import ProcessingStep\n",
    "\n",
    "docker_name = \"sagemaker-preprocessing\"\n",
    "account = sagemaker_session.boto_session.client(\"sts\").get_caller_identity()[\"Account\"]\n",
    "region = sagemaker_session.boto_session.region_name\n",
    "image = \"{}.dkr.ecr.{}.amazonaws.com/{}:latest\".format(account, region, docker_name)\n",
    "print(image)\n",
    "script_processor = ScriptProcessor(image_uri=image,\n",
    "                role=role,\n",
    "                instance_count=1,\n",
    "                instance_type='ml.m5.xlarge',\n",
    "                command=['python3'],\n",
    "                sagemaker_session=pipeline_session)\n",
    "\n",
    "\n",
    "processor_args=script_processor.run(code='scripts/preprocessing.py',\n",
    "                    arguments = [\"--bucket\",bucket,'--region',region])\n",
    "\n",
    "\n",
    "\n",
    "step_process = ProcessingStep(\n",
    "    name=\"PreprocessData\",\n",
    "    step_args=processor_args,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "29116067",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting scripts/train.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile scripts/train.py\n",
    "\n",
    "import argparse\n",
    "import os\n",
    "import pandas as pd\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.externals import joblib\n",
    "from sklearn.preprocessing import Normalizer\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    parser = argparse.ArgumentParser()\n",
    "    \n",
    "    # Hyperparameters are described here. In this simple example we are just including one hyperparameter.\n",
    "    parser.add_argument('--n_clusters', type=int, default=2)\n",
    "    parser.add_argument('--random_state', type=int, default=0)\n",
    "    \n",
    "    # Sagemaker specific arguments. Defaults are set in the environment variables.\n",
    "    parser.add_argument('--output-data-dir', type=str, default=os.environ['SM_OUTPUT_DATA_DIR'])\n",
    "    parser.add_argument('--model-dir', type=str, default=os.environ['SM_MODEL_DIR'])\n",
    "    parser.add_argument('--training', type=str, default=os.environ['SM_CHANNEL_TRAINING'])\n",
    "\n",
    "    args = parser.parse_args()\n",
    "    \n",
    "    input_files = [ os.path.join(args.training, file) for file in os.listdir(args.training) ]\n",
    "    if len(input_files) == 0:\n",
    "        raise ValueError(('There are no files in {}.\\n' +\n",
    "                          'This usually indicates that the channel ({}) was incorrectly specified,\\n' +\n",
    "                          'the data specification in S3 was incorrectly specified or the role specified\\n' +\n",
    "                          'does not have permission to access the data.').format(args.training, \"train\"))\n",
    "    \n",
    "    raw_data = [ pd.read_csv(file) for file in input_files ]\n",
    "    train_data = pd.concat(raw_data)\n",
    "    print(train_data.shape)\n",
    "    kmeans = KMeans(n_clusters=2,random_state=0).fit(train_data)\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Print the coefficients of the trained classifier, and save the coefficients\n",
    "    joblib.dump(kmeans, os.path.join(args.model_dir, \"kmeansmodel.joblib\"))\n",
    "    \n",
    "def model_fn(model_dir):\n",
    "    \"\"\"Deserialized and return fitted model\n",
    "\n",
    "    Note that this should have the same name as the serialized model in the main method\n",
    "    \"\"\"\n",
    "    kmeans = joblib.load(os.path.join(model_dir, \"kmeansmodel.joblib\"))\n",
    "    return kmeans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "75cf0d7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sagemaker.workflow.steps import TrainingStep\n",
    "\n",
    "model_path= f\"s3://{bucket}/{prefix}/model/\"\n",
    "\n",
    "\n",
    "sklearn = SKLearn(\n",
    "    source_dir='scripts',\n",
    "    entry_point='train.py',\n",
    "    instance_type=\"ml.m4.xlarge\",\n",
    "    role = role,\n",
    "    sagemaker_session=pipeline_session,\n",
    "    framework_version='0.20.0',\n",
    "    output_path=model_path,\n",
    "    hyperparameters={'n_clusters': 2, 'random_state':0})\n",
    "\n",
    "train_args=sklearn.fit({'training': 's3://{}/{}/prediction_data.csv'.format(bucket,prefix)})\n",
    "step_train_model = TrainingStep(name=\"TrainModel\", step_args=train_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c342e009",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sagemaker.model import Model\n",
    "from sagemaker.sklearn.model import SKLearnModel\n",
    "from sagemaker.model import Model\n",
    "from sagemaker.workflow.model_step import ModelStep\n",
    "\n",
    "\n",
    "\n",
    "clustering_model = SKLearnModel(\n",
    "    model_data=step_train_model.properties.ModelArtifacts.S3ModelArtifacts,\n",
    "    role=role,\n",
    "    sagemaker_session=pipeline_session,\n",
    "    entry_point=\"scripts/train.py\",\n",
    "    framework_version='0.20.0'\n",
    ")\n",
    "\n",
    "\n",
    "step_create_model = ModelStep(\n",
    "    name=\"clustering_model\",\n",
    "    step_args=clustering_model.create(instance_type=\"ml.m5.large\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6d6069a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from sagemaker.transformer import Transformer\n",
    "# from sagemaker.workflow.steps import TransformStep\n",
    "\n",
    "# transformer = Transformer(\n",
    "#     model_name=step_create_model.properties.ModelName,\n",
    "#     instance_type=\"ml.m5.xlarge\",\n",
    "#     instance_count=1,\n",
    "#     output_path=f\"s3://{bucket}/{prefix}/transform-output\",\n",
    "#     sagemaker_session=pipeline_session\n",
    "# )\n",
    "\n",
    "# prediction_data_path=\"s3://{}/{}/prediction_data.csv\".format(bucket,prefix)\n",
    "\n",
    "# transform_args=transformer.transform(prediction_data_path, content_type='text/csv', split_type = 'Line')\n",
    "# step_transform = TransformStep(name=\"clustering_model\",step_args= transform_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "85f7ef42",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sagemaker.workflow.pipeline import Pipeline\n",
    "\n",
    "# Create a Sagemaker Pipeline.\n",
    "# Each parameter for the pipeline must be set as a parameter explicitly when the pipeline is created.\n",
    "# Also pass in each of the steps created above.\n",
    "# Note that the order of execution is determined from each step's dependencies on other steps,\n",
    "# not on the order they are passed in below.\n",
    "pipeline = Pipeline(\n",
    "    name=\"adverse-drug-reaction\",\n",
    "    steps=[step_process, step_train_model, step_create_model]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bc9ec38e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Version': '2020-12-01',\n",
       " 'Metadata': {},\n",
       " 'Parameters': [],\n",
       " 'PipelineExperimentConfig': {'ExperimentName': {'Get': 'Execution.PipelineName'},\n",
       "  'TrialName': {'Get': 'Execution.PipelineExecutionId'}},\n",
       " 'Steps': [{'Name': 'PreprocessData',\n",
       "   'Type': 'Processing',\n",
       "   'Arguments': {'ProcessingResources': {'ClusterConfig': {'InstanceType': 'ml.m5.xlarge',\n",
       "      'InstanceCount': 1,\n",
       "      'VolumeSizeInGB': 30}},\n",
       "    'AppSpecification': {'ImageUri': '485822383573.dkr.ecr.us-east-1.amazonaws.com/sagemaker-preprocessing:latest',\n",
       "     'ContainerArguments': ['--bucket',\n",
       "      'sagemaker-us-east-1-485822383573',\n",
       "      '--region',\n",
       "      'us-east-1'],\n",
       "     'ContainerEntrypoint': ['python3',\n",
       "      '/opt/ml/processing/input/code/preprocessing.py']},\n",
       "    'RoleArn': 'arn:aws:iam::485822383573:role/service-role/AmazonSageMaker-ExecutionRole-20220426T122295',\n",
       "    'ProcessingInputs': [{'InputName': 'code',\n",
       "      'AppManaged': False,\n",
       "      'S3Input': {'S3Uri': 's3://sagemaker-us-east-1-485822383573/sagemaker-preprocessing-2022-08-13-05-03-07-681/input/code/preprocessing.py',\n",
       "       'LocalPath': '/opt/ml/processing/input/code',\n",
       "       'S3DataType': 'S3Prefix',\n",
       "       'S3InputMode': 'File',\n",
       "       'S3DataDistributionType': 'FullyReplicated',\n",
       "       'S3CompressionType': 'None'}}]}},\n",
       "  {'Name': 'TrainModel',\n",
       "   'Type': 'Training',\n",
       "   'Arguments': {'AlgorithmSpecification': {'TrainingInputMode': 'File',\n",
       "     'TrainingImage': '683313688378.dkr.ecr.us-east-1.amazonaws.com/sagemaker-scikit-learn:0.20.0-cpu-py3'},\n",
       "    'OutputDataConfig': {'S3OutputPath': 's3://sagemaker-us-east-1-485822383573/chapter9/data/model/'},\n",
       "    'StoppingCondition': {'MaxRuntimeInSeconds': 86400},\n",
       "    'ResourceConfig': {'VolumeSizeInGB': 30,\n",
       "     'InstanceCount': 1,\n",
       "     'InstanceType': 'ml.m4.xlarge'},\n",
       "    'RoleArn': 'arn:aws:iam::485822383573:role/service-role/AmazonSageMaker-ExecutionRole-20220426T122295',\n",
       "    'InputDataConfig': [{'DataSource': {'S3DataSource': {'S3DataType': 'S3Prefix',\n",
       "        'S3Uri': 's3://sagemaker-us-east-1-485822383573/chapter9/data/prediction_data.csv',\n",
       "        'S3DataDistributionType': 'FullyReplicated'}},\n",
       "      'ChannelName': 'training'}],\n",
       "    'HyperParameters': {'n_clusters': '2',\n",
       "     'random_state': '0',\n",
       "     'sagemaker_submit_directory': '\"s3://sagemaker-us-east-1-485822383573/sagemaker-scikit-learn-2022-08-13-05-03-07-983/source/sourcedir.tar.gz\"',\n",
       "     'sagemaker_program': '\"train.py\"',\n",
       "     'sagemaker_container_log_level': '20',\n",
       "     'sagemaker_region': '\"us-east-1\"'},\n",
       "    'DebugHookConfig': {'S3OutputPath': 's3://sagemaker-us-east-1-485822383573/chapter9/data/model/',\n",
       "     'CollectionConfigurations': []},\n",
       "    'ProfilerRuleConfigurations': [{'RuleConfigurationName': 'ProfilerReport-1660366988',\n",
       "      'RuleEvaluatorImage': '503895931360.dkr.ecr.us-east-1.amazonaws.com/sagemaker-debugger-rules:latest',\n",
       "      'RuleParameters': {'rule_to_invoke': 'ProfilerReport'}}],\n",
       "    'ProfilerConfig': {'S3OutputPath': 's3://sagemaker-us-east-1-485822383573/chapter9/data/model/'}}},\n",
       "  {'Name': 'clustering_model-CreateModel',\n",
       "   'Type': 'Model',\n",
       "   'Arguments': {'ExecutionRoleArn': 'arn:aws:iam::485822383573:role/service-role/AmazonSageMaker-ExecutionRole-20220426T122295',\n",
       "    'PrimaryContainer': {'Image': '683313688378.dkr.ecr.us-east-1.amazonaws.com/sagemaker-scikit-learn:0.20.0-cpu-py3',\n",
       "     'Environment': {'SAGEMAKER_PROGRAM': 'train.py',\n",
       "      'SAGEMAKER_SUBMIT_DIRECTORY': 's3://sagemaker-us-east-1-485822383573/sagemaker-scikit-learn-2022-08-13-05-03-08-148/sourcedir.tar.gz',\n",
       "      'SAGEMAKER_CONTAINER_LOG_LEVEL': '20',\n",
       "      'SAGEMAKER_REGION': 'us-east-1'},\n",
       "     'ModelDataUrl': {'Get': 'Steps.TrainModel.ModelArtifacts.S3ModelArtifacts'}}}}]}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "definition = json.loads(pipeline.definition())\n",
    "definition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "79cd0fd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'PipelineArn': 'arn:aws:sagemaker:us-east-1:485822383573:pipeline/adverse-drug-reaction',\n",
       " 'ResponseMetadata': {'RequestId': '69d02f24-ea53-4f9f-9868-528f05522fa3',\n",
       "  'HTTPStatusCode': 200,\n",
       "  'HTTPHeaders': {'x-amzn-requestid': '69d02f24-ea53-4f9f-9868-528f05522fa3',\n",
       "   'content-type': 'application/x-amz-json-1.1',\n",
       "   'content-length': '89',\n",
       "   'date': 'Sat, 13 Aug 2022 05:05:05 GMT'},\n",
       "  'RetryAttempts': 0}}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "pipeline.upsert(role_arn=role)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0921e803",
   "metadata": {},
   "outputs": [],
   "source": [
    "execution = pipeline.start()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "de7e5820",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'PipelineArn': 'arn:aws:sagemaker:us-east-1:485822383573:pipeline/adverse-drug-reaction',\n",
       " 'PipelineExecutionArn': 'arn:aws:sagemaker:us-east-1:485822383573:pipeline/adverse-drug-reaction/execution/9froltkry8hx',\n",
       " 'PipelineExecutionDisplayName': 'execution-1660367106574',\n",
       " 'PipelineExecutionStatus': 'Executing',\n",
       " 'PipelineExperimentConfig': {'ExperimentName': 'adverse-drug-reaction',\n",
       "  'TrialName': '9froltkry8hx'},\n",
       " 'CreationTime': datetime.datetime(2022, 8, 13, 5, 5, 6, 395000, tzinfo=tzlocal()),\n",
       " 'LastModifiedTime': datetime.datetime(2022, 8, 13, 5, 5, 6, 395000, tzinfo=tzlocal()),\n",
       " 'CreatedBy': {},\n",
       " 'LastModifiedBy': {},\n",
       " 'ResponseMetadata': {'RequestId': '78271b7d-fa2a-430a-8a75-c90d3049cb6f',\n",
       "  'HTTPStatusCode': 200,\n",
       "  'HTTPHeaders': {'x-amzn-requestid': '78271b7d-fa2a-430a-8a75-c90d3049cb6f',\n",
       "   'content-type': 'application/x-amz-json-1.1',\n",
       "   'content-length': '504',\n",
       "   'date': 'Sat, 13 Aug 2022 05:05:10 GMT'},\n",
       "  'RetryAttempts': 0}}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "execution.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d3b6a20f",
   "metadata": {},
   "outputs": [],
   "source": [
    "execution.wait()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6c7fa3b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'StepName': 'clustering_model-CreateModel',\n",
       "  'StartTime': datetime.datetime(2022, 8, 13, 5, 9, 3, 562000, tzinfo=tzlocal()),\n",
       "  'EndTime': datetime.datetime(2022, 8, 13, 5, 9, 4, 641000, tzinfo=tzlocal()),\n",
       "  'StepStatus': 'Succeeded',\n",
       "  'AttemptCount': 0,\n",
       "  'Metadata': {'Model': {'Arn': 'arn:aws:sagemaker:us-east-1:485822383573:model/pipelines-9froltkry8hx-clustering-model-cre-ghw6cyinbs'}}},\n",
       " {'StepName': 'TrainModel',\n",
       "  'StartTime': datetime.datetime(2022, 8, 13, 5, 5, 7, 761000, tzinfo=tzlocal()),\n",
       "  'EndTime': datetime.datetime(2022, 8, 13, 5, 9, 3, 106000, tzinfo=tzlocal()),\n",
       "  'StepStatus': 'Succeeded',\n",
       "  'AttemptCount': 0,\n",
       "  'Metadata': {'TrainingJob': {'Arn': 'arn:aws:sagemaker:us-east-1:485822383573:training-job/pipelines-9froltkry8hx-trainmodel-qjhvj4ch4m'}}},\n",
       " {'StepName': 'PreprocessData',\n",
       "  'StartTime': datetime.datetime(2022, 8, 13, 5, 5, 7, 761000, tzinfo=tzlocal()),\n",
       "  'EndTime': datetime.datetime(2022, 8, 13, 5, 9, 28, 886000, tzinfo=tzlocal()),\n",
       "  'StepStatus': 'Succeeded',\n",
       "  'AttemptCount': 0,\n",
       "  'Metadata': {'ProcessingJob': {'Arn': 'arn:aws:sagemaker:us-east-1:485822383573:processing-job/pipelines-9froltkry8hx-preprocessdata-azv4gnc61f'}}}]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "execution.list_steps()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e619ac36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "...........................\n",
      "\u001b[34mProcessing /opt/ml/code\n",
      "  DEPRECATION: A future pip version will change local packages to be built in-place without first copying to a temporary directory. We recommend you use --use-feature=in-tree-build to test your packages with this new behavior before it becomes the default.\n",
      "   pip 21.3 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/7555.\u001b[0m\n",
      "\u001b[34mBuilding wheels for collected packages: train\n",
      "  Building wheel for train (setup.py): started\n",
      "  Building wheel for train (setup.py): finished with status 'done'\n",
      "  Created wheel for train: filename=train-1.0.0-py2.py3-none-any.whl size=4394 sha256=cda5e55692be74bb20f7b02c10d9e3a490cfaf0e31ee1d2919c4af0fa57c3705\n",
      "  Stored in directory: /tmp/pip-ephem-wheel-cache-_nmp60pe/wheels/3e/0f/51/2f1df833dd0412c1bc2f5ee56baac195b5be563353d111dca6\u001b[0m\n",
      "\u001b[34mSuccessfully built train\u001b[0m\n",
      "\u001b[34mInstalling collected packages: train\u001b[0m\n",
      "\u001b[34mSuccessfully installed train-1.0.0\u001b[0m\n",
      "\u001b[34mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\n",
      "\u001b[34m/miniconda3/lib/python3.7/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses\n",
      "  import imp\u001b[0m\n",
      "\u001b[34m[2022-08-13 05:32:06 +0000] [38] [INFO] Starting gunicorn 20.1.0\u001b[0m\n",
      "\u001b[34m[2022-08-13 05:32:06 +0000] [38] [INFO] Listening at: unix:/tmp/gunicorn.sock (38)\u001b[0m\n",
      "\u001b[34m[2022-08-13 05:32:06 +0000] [38] [INFO] Using worker: gevent\u001b[0m\n",
      "\u001b[34m[2022-08-13 05:32:06 +0000] [41] [INFO] Booting worker with pid: 41\u001b[0m\n",
      "\u001b[34m[2022-08-13 05:32:06 +0000] [42] [INFO] Booting worker with pid: 42\u001b[0m\n",
      "\u001b[34m[2022-08-13 05:32:06 +0000] [43] [INFO] Booting worker with pid: 43\u001b[0m\n",
      "\u001b[34m[2022-08-13 05:32:06 +0000] [47] [INFO] Booting worker with pid: 47\u001b[0m\n",
      "\u001b[34m2022-08-13 05:32:09,011 INFO - sagemaker-containers - No GPUs detected (normal if no gpus installed)\u001b[0m\n",
      "\u001b[34m/miniconda3/lib/python3.7/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses\n",
      "  import imp\u001b[0m\n",
      "\u001b[34m169.254.255.130 - - [13/Aug/2022:05:32:09 +0000] \"GET /ping HTTP/1.1\" 200 0 \"-\" \"Go-http-client/1.1\"\u001b[0m\n",
      "\u001b[34m169.254.255.130 - - [13/Aug/2022:05:32:09 +0000] \"GET /execution-parameters HTTP/1.1\" 404 232 \"-\" \"Go-http-client/1.1\"\u001b[0m\n",
      "\u001b[34m2022-08-13 05:32:09,705 ERROR - train - Exception on /invocations [POST]\u001b[0m\n",
      "\u001b[34mTraceback (most recent call last):\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 93, in wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_sklearn_container/serving.py\", line 66, in default_predict_fn\n",
      "    output = model.predict(input_data)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 1073, in predict\n",
      "    X = self._check_test_data(X)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 936, in _check_test_data\n",
      "    n_features, expected_n_features))\u001b[0m\n",
      "\u001b[34mValueError: Incorrect number of features. Got 220 features, expected 216\u001b[0m\n",
      "\u001b[34mDuring handling of the above exception, another exception occurred:\u001b[0m\n",
      "\u001b[34mTraceback (most recent call last):\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_transformer.py\", line 200, in transform\n",
      "    self._model, request.content, request.content_type, request.accept\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_transformer.py\", line 231, in _default_transform_fn\n",
      "    prediction = self._predict_fn(data, model)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 95, in wrapper\n",
      "    six.reraise(error_class, error_class(e), sys.exc_info()[2])\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/six.py\", line 718, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 93, in wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_sklearn_container/serving.py\", line 66, in default_predict_fn\n",
      "    output = model.predict(input_data)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 1073, in predict\n",
      "    X = self._check_test_data(X)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 936, in _check_test_data\n",
      "    n_features, expected_n_features))\u001b[0m\n",
      "\u001b[34msagemaker_containers._errors.ClientError: Incorrect number of features. Got 220 features, expected 216\u001b[0m\n",
      "\u001b[34m169.254.255.130 - - [13/Aug/2022:05:32:09 +0000] \"POST /invocations HTTP/1.1\" 500 290 \"-\" \"Go-http-client/1.1\"\u001b[0m\n",
      "\u001b[34m2022-08-13 05:32:09,725 INFO - sagemaker-containers - No GPUs detected (normal if no gpus installed)\u001b[0m\n",
      "\u001b[34m/miniconda3/lib/python3.7/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses\n",
      "  import imp\u001b[0m\n",
      "\u001b[34m2022-08-13 05:32:10,276 ERROR - train - Exception on /invocations [POST]\u001b[0m\n",
      "\u001b[34mTraceback (most recent call last):\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 93, in wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_sklearn_container/serving.py\", line 66, in default_predict_fn\n",
      "    output = model.predict(input_data)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 1073, in predict\n",
      "    X = self._check_test_data(X)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 936, in _check_test_data\n",
      "    n_features, expected_n_features))\u001b[0m\n",
      "\u001b[34mValueError: Incorrect number of features. Got 220 features, expected 216\u001b[0m\n",
      "\u001b[34mDuring handling of the above exception, another exception occurred:\u001b[0m\n",
      "\u001b[34mTraceback (most recent call last):\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_transformer.py\", line 200, in transform\n",
      "    self._model, request.content, request.content_type, request.accept\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_transformer.py\", line 231, in _default_transform_fn\n",
      "    prediction = self._predict_fn(data, model)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 95, in wrapper\n",
      "    six.reraise(error_class, error_class(e), sys.exc_info()[2])\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/six.py\", line 718, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 93, in wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_sklearn_container/serving.py\", line 66, in default_predict_fn\n",
      "    output = model.predict(input_data)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 1073, in predict\n",
      "    X = self._check_test_data(X)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 936, in _check_test_data\n",
      "    n_features, expected_n_features))\u001b[0m\n",
      "\u001b[34msagemaker_containers._errors.ClientError: Incorrect number of features. Got 220 features, expected 216\u001b[0m\n",
      "\u001b[34m169.254.255.130 - - [13/Aug/2022:05:32:10 +0000] \"POST /invocations HTTP/1.1\" 500 290 \"-\" \"Go-http-client/1.1\"\u001b[0m\n",
      "\u001b[34m2022-08-13 05:32:10,295 INFO - sagemaker-containers - No GPUs detected (normal if no gpus installed)\u001b[0m\n",
      "\u001b[34m/miniconda3/lib/python3.7/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses\n",
      "  import imp\u001b[0m\n",
      "\u001b[34m2022-08-13 05:32:10,827 ERROR - train - Exception on /invocations [POST]\u001b[0m\n",
      "\u001b[34mTraceback (most recent call last):\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 93, in wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_sklearn_container/serving.py\", line 66, in default_predict_fn\n",
      "    output = model.predict(input_data)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 1073, in predict\n",
      "    X = self._check_test_data(X)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 936, in _check_test_data\n",
      "    n_features, expected_n_features))\u001b[0m\n",
      "\u001b[34mValueError: Incorrect number of features. Got 220 features, expected 216\u001b[0m\n",
      "\u001b[34mDuring handling of the above exception, another exception occurred:\u001b[0m\n",
      "\u001b[34mTraceback (most recent call last):\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_transformer.py\", line 200, in transform\n",
      "    self._model, request.content, request.content_type, request.accept\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_transformer.py\", line 231, in _default_transform_fn\n",
      "    prediction = self._predict_fn(data, model)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 95, in wrapper\n",
      "    six.reraise(error_class, error_class(e), sys.exc_info()[2])\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/six.py\", line 718, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 93, in wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_sklearn_container/serving.py\", line 66, in default_predict_fn\n",
      "    output = model.predict(input_data)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 1073, in predict\n",
      "    X = self._check_test_data(X)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 936, in _check_test_data\n",
      "    n_features, expected_n_features))\u001b[0m\n",
      "\u001b[34msagemaker_containers._errors.ClientError: Incorrect number of features. Got 220 features, expected 216\u001b[0m\n",
      "\u001b[34m169.254.255.130 - - [13/Aug/2022:05:32:10 +0000] \"POST /invocations HTTP/1.1\" 500 290 \"-\" \"Go-http-client/1.1\"\u001b[0m\n",
      "\u001b[34m2022-08-13 05:32:10,865 ERROR - train - Exception on /invocations [POST]\u001b[0m\n",
      "\u001b[34mTraceback (most recent call last):\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 93, in wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_sklearn_container/serving.py\", line 66, in default_predict_fn\n",
      "    output = model.predict(input_data)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 1073, in predict\n",
      "    X = self._check_test_data(X)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 936, in _check_test_data\n",
      "    n_features, expected_n_features))\u001b[0m\n",
      "\u001b[34mValueError: Incorrect number of features. Got 220 features, expected 216\u001b[0m\n",
      "\u001b[34mDuring handling of the above exception, another exception occurred:\u001b[0m\n",
      "\u001b[34mTraceback (most recent call last):\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/flask/app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_transformer.py\", line 200, in transform\n",
      "    self._model, request.content, request.content_type, request.accept\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_transformer.py\", line 231, in _default_transform_fn\n",
      "    prediction = self._predict_fn(data, model)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 95, in wrapper\n",
      "    six.reraise(error_class, error_class(e), sys.exc_info()[2])\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/six.py\", line 718, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_containers/_functions.py\", line 93, in wrapper\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sagemaker_sklearn_container/serving.py\", line 66, in default_predict_fn\n",
      "    output = model.predict(input_data)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 1073, in predict\n",
      "    X = self._check_test_data(X)\n",
      "  File \"/miniconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py\", line 936, in _check_test_data\n",
      "    n_features, expected_n_features))\u001b[0m\n",
      "\u001b[34msagemaker_containers._errors.ClientError: Incorrect number of features. Got 220 features, expected 216\u001b[0m\n",
      "\u001b[34m169.254.255.130 - - [13/Aug/2022:05:32:10 +0000] \"POST /invocations HTTP/1.1\" 500 290 \"-\" \"Go-http-client/1.1\"\u001b[0m\n",
      "\u001b[32m2022-08-13T05:32:09.589:[sagemaker logs]: MaxConcurrentTransforms=1, MaxPayloadInMB=6, BatchStrategy=MULTI_RECORD\u001b[0m\n",
      "\u001b[32m2022-08-13T05:32:10.868:[sagemaker logs]: sagemaker-us-east-1-485822383573/chapter9/data/prediction_data.csv: Bad HTTP status received from algorithm: 500\u001b[0m\n",
      "\u001b[32m2022-08-13T05:32:10.869:[sagemaker logs]: sagemaker-us-east-1-485822383573/chapter9/data/prediction_data.csv: \u001b[0m\n",
      "\u001b[32m2022-08-13T05:32:10.869:[sagemaker logs]: sagemaker-us-east-1-485822383573/chapter9/data/prediction_data.csv: Message:\u001b[0m\n",
      "\u001b[32m2022-08-13T05:32:10.869:[sagemaker logs]: sagemaker-us-east-1-485822383573/chapter9/data/prediction_data.csv: <!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 3.2 Final//EN\">\u001b[0m\n",
      "\u001b[32m2022-08-13T05:32:10.869:[sagemaker logs]: sagemaker-us-east-1-485822383573/chapter9/data/prediction_data.csv: <title>500 Internal Server Error</title>\u001b[0m\n",
      "\u001b[32m2022-08-13T05:32:10.869:[sagemaker logs]: sagemaker-us-east-1-485822383573/chapter9/data/prediction_data.csv: <h1>Internal Server Error</h1>\u001b[0m\n",
      "\u001b[32m2022-08-13T05:32:10.869:[sagemaker logs]: sagemaker-us-east-1-485822383573/chapter9/data/prediction_data.csv: <p>The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.</p>\u001b[0m\n"
     ]
    },
    {
     "ename": "UnexpectedStatusException",
     "evalue": "Error for Transform job sagemaker-scikit-learn-2022-08-13-05-27-44-089: Failed. Reason: AlgorithmError: See job logs for more information",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnexpectedStatusException\u001b[0m                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-33-ab0c4c2dbf34>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0mprediction_data_path\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"s3://{}/{}/prediction_data.csv\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbucket\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mprefix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m \u001b[0mtransformer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtransform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprediction_data_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontent_type\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'text/csv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msplit_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'Line'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/python3/lib/python3.6/site-packages/sagemaker/workflow/pipeline_context.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    208\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself_instance\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msagemaker_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontext\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    209\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 210\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mrun_func\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    211\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    212\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/python3/lib/python3.6/site-packages/sagemaker/transformer.py\u001b[0m in \u001b[0;36mtransform\u001b[0;34m(self, data, data_type, content_type, compression_type, split_type, job_name, input_filter, output_filter, join_source, experiment_config, model_client_config, wait, logs)\u001b[0m\n\u001b[1;32m    235\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    236\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 237\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlatest_transform_job\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlogs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlogs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    238\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    239\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdelete_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/python3/lib/python3.6/site-packages/sagemaker/transformer.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, logs)\u001b[0m\n\u001b[1;32m    432\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlogs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    433\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlogs\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 434\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msagemaker_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlogs_for_transform_job\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjob_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    435\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    436\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msagemaker_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait_for_transform_job\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjob_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/python3/lib/python3.6/site-packages/sagemaker/session.py\u001b[0m in \u001b[0;36mlogs_for_transform_job\u001b[0;34m(self, job_name, wait, poll)\u001b[0m\n\u001b[1;32m   4019\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4020\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4021\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_check_job_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjob_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdescription\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"TransformJobStatus\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4022\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mdot\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4023\u001b[0m                 \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/python3/lib/python3.6/site-packages/sagemaker/session.py\u001b[0m in \u001b[0;36m_check_job_status\u001b[0;34m(self, job, desc, status_key_name)\u001b[0m\n\u001b[1;32m   3390\u001b[0m                 \u001b[0mmessage\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3391\u001b[0m                 \u001b[0mallowed_statuses\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Completed\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Stopped\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3392\u001b[0;31m                 \u001b[0mactual_status\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstatus\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3393\u001b[0m             )\n\u001b[1;32m   3394\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnexpectedStatusException\u001b[0m: Error for Transform job sagemaker-scikit-learn-2022-08-13-05-27-44-089: Failed. Reason: AlgorithmError: See job logs for more information"
     ]
    }
   ],
   "source": [
    "from sagemaker.transformer import Transformer\n",
    "from sagemaker.workflow.steps import TransformStep\n",
    "\n",
    "transformer = Transformer(\n",
    "    model_name=\"pipelines-9froltkry8hx-clustering-model-Cre-gHw6cYInBS\",\n",
    "    instance_type=\"ml.m5.xlarge\",\n",
    "    instance_count=1,\n",
    "    output_path=f\"s3://{bucket}/{prefix}/transform-output\",\n",
    "    sagemaker_session=sagemaker_session\n",
    ")\n",
    "\n",
    "prediction_data_path=\"s3://{}/{}/prediction_data.csv\".format(bucket,prefix)\n",
    "transformer.transform(prediction_data_path, content_type='text/csv', split_type = 'Line')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a4e31024",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sagemaker.workflow.properties.Properties at 0x7f58601377b8>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transform_output = transformer.output_path\n",
    "# !mkdir -p cluster_transformer/output\n",
    "# !aws s3 cp --recursive $batch_output/ cluster_transformer/output/\n",
    "# # Head to see what the batch output looks like\n",
    "# !head cluster_transformer/output/*\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "18cc769e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pipeline=execution.list_steps()[0]['Metadata']['Model']['Arn'].split('/')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a09e804d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'pipelines-9froltkry8hx-clustering-model-cre-ghw6cyinbs'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "execution.list_steps()[0]['Metadata']['Model']['Arn'].split('/')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a85ae012",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_path=\"s3://{bucket}/{prefix}/model/{pipeline}/\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
